src/backend/distributed/utils/enable_ssl.c

```diff
@@ -0,0 +1,457 @@
+/*-------------------------------------------------------------------------
+ *
+ * enable_ssl.c
+ *    UDF and Utilities for enabling ssl during citus setup
+ *
+ * Copyright (c) 2018, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "distributed/connection_management.h"
+#include "distributed/worker_protocol.h"
+#include "fmgr.h"
+#include "libpq/libpq.h"
+#include "nodes/parsenodes.h"
+#include "postmaster/postmaster.h"
+#include "utils/guc.h"
+
+#ifdef USE_OPENSSL
+#include "openssl/dsa.h"
+#include "openssl/err.h"
+#include "openssl/pem.h"
+#include "openssl/rsa.h"
+#include "openssl/ssl.h"
+#include "openssl/x509.h"
+#endif
+
+
+#define DirectFunctionCall0(func) \
+	DirectFunctionCall0Coll(func, InvalidOid)
+
+#define ENABLE_SSL_QUERY "ALTER SYSTEM SET ssl TO on;"
+#define RESET_CITUS_NODE_CONNINFO \
+	"ALTER SYSTEM SET citus.node_conninfo TO 'sslmode=prefer';"
+
+#define CITUS_AUTO_SSL_COMMON_NAME "citus-auto-ssl"
+#define X509_SUBJECT_COMMON_NAME "CN"
+
+
+/* forward declaration of helper functions */
+static Datum DirectFunctionCall0Coll(PGFunction func, Oid collation);
+
+/* use pg's implementation that is not exposed in a header file */
+extern Datum pg_reload_conf(PG_FUNCTION_ARGS);
+
+
+#ifdef USE_SSL
+
+/* forward declaration of functions used when compiled with ssl */
+static void EnsureReleaseOpenSSLResource(MemoryContextCallbackFunction free, void *arg);
+static bool ShouldUseAutoSSL(void);
+static bool CreateCertificatesWhenNeeded(void);
+static EVP_PKEY * GeneratePrivateKey(void);
+static X509 * CreateCertificate(EVP_PKEY *privateKey);
+static bool StoreCertificate(EVP_PKEY *privateKey, X509 *certificate);
+#endif /* USE_SSL */
+
+
+PG_FUNCTION_INFO_V1(citus_setup_ssl);
+PG_FUNCTION_INFO_V1(citus_reset_default_for_node_conninfo);
+
+
+/*
+ * citus_setup_ssl is called during the first creation of a citus extension. It configures
+ * postgres to use ssl if not already on. During this process it will create certificates
+ * if they are not already installed in the configured location.
+ */
+Datum
+citus_setup_ssl(PG_FUNCTION_ARGS)
+{
+#ifndef USE_SSL
+	ereport(WARNING, (errmsg("can not setup ssl on postgres that is not compiled with "
```

Comment:
> For instance our windows builds on appveyor do not have postgres compiled with ssl, so that would break those builds.

So, also today, we decided to kill the appveyor stuff. It makes our PR dashboard look bad to outsiders who won't know why every PR has an ❌ next to it.

If we do the package-based approach to certificate generation, do we still need to link to openssl? No, right?

Reply:
Depends,

Currently we check certificates existing via an openssl library function. We can decide to only stat the file instead. Then we could remove openssl.